tools/nngcat/nngcat.c

```diff
@@ -309,28 +310,49 @@ static void
 loadfile(const char *path, void **datap, size_t *lenp)
 {
 	FILE * f;
+	size_t total_read = 0;
+	size_t allocation_size = BUFSIZ;
 	char * fdata;
-	size_t len;
 
-	if ((f = fopen(path, "r")) == NULL) {
-		fatal("Cannot open file %s: %s", path, strerror(errno));
-	}
-	if (fseek(f, 0, SEEK_END) != 0) {
-		fatal("Cannot seek to end of file: %s", strerror(errno));
+	if (strcmp(path, "-") == 0) {
+		f = stdin;
+	} else {
+		if ((f = fopen(path, "rb")) == NULL) {
+			fatal("Cannot open file %s: %s", path, strerror(errno));
+		}
 	}
-	len = ftell(f);
-	(void) fseek(f, 0, SEEK_SET);
-	if ((fdata = malloc(len + 1)) == NULL) {
+
+	if ((fdata = malloc(allocation_size + 1)) == NULL) {
 		fatal("Out of memory.");
 	}
-	fdata[len] = '\0';
 
-	if (fread(fdata, 1, len, f) != len) {
-		fatal("Read file %s failed: %s", path, strerror(errno));
+	while (1) {
+		total_read += fread(fdata + total_read, 1, allocation_size - total_read, f);
+		if (ferror(f)) {
+			if (errno == EINTR) {
+				continue;
+			}
+			fatal("Read from %s failed: %s", path, strerror(errno));
+		}
+		if (feof(f)) {
+			break;
+		}
+		if (total_read == allocation_size) {
+			if (allocation_size > SIZE_MAX / 2) {
+				fatal("Out of memory.");
+			}
+			allocation_size *= 2;
+			if ((fdata = realloc(fdata, allocation_size + 1)) == NULL) {
```

Comment:
Codacy complains that this is a memory leak, but since we just `fatal` on realloc failure anyway I don't think it matters too much. Happy to fix it if you like though.

Reply:
I'd like to fix it please.  I agree it's probably benign, but some users get unhappy when automated code analysis tools trigger complaints.

Reply:
Should be fixed now